File: Hw_2/AVS_Rakhmatullin207_hw2/functional.cpp
//------------------------------------------------------------------------------
// functional.cpp - содержит функции обработки функционального языка
//------------------------------------------------------------------------------

#include "functional.h"

//------------------------------------------------------------------------------
// Ввод параметров функционального языка из файла
void Functional::In(FILE *input_file) {
    int inputType;
    fscanf(input_file, "%d", &inputType);
    switch (inputType) {
        case 1:
            type = STRICT;
        case 2:
            type = DYNAMIC;
        default:
            type = STRICT;
    }
    int is_lazy;
    fscanf(input_file, "%d", &is_lazy);
    lazy = is_lazy == 1;
}

// Случайный ввод параметров функционального языка
void Functional::InRnd() {
    switch (rnd2.Get()) {
        case 1:
            type = STRICT;
        case 2:
            type = DYNAMIC;
        default:
            type = STRICT;
    }
    lazy = rnd2.Get() == 1;
}

//------------------------------------------------------------------------------
// Вывод параметров функционального языка в форматируемый поток
void Functional::Out(FILE *output_file) {
    char typization[8];
    switch (type) {
        case STRICT:
            strcpy(typization, "Strict");
        case DYNAMIC:
            strcpy(typization, "Dynamic");
        default:
            strcpy(typization, "Strict");
    }
    char is_lazy[4];
    strcpy(is_lazy, lazy ? "yes" : "no");
    fprintf(output_file, "Type: functional\tLazy: %s\tTypization - %s\n", is_lazy, typization);
    fprintf(output_file, "Year of founding: %d", founding_year);
    fprintf(output_file, "\tPopularity: %lf %%\n", popularity);
    fprintf(output_file, "Ratio of founding year to name length: %lf\n", Proportion());
}

File: Hw_3/AVS_Rakhmatullin_hw3/functional.py
import enum
from checkers import IsFactor


# Типизация языка
class Typization(enum.Enum):
    strict = 1
    dynamic = 2


# Функциональный язык
class Functional:
    def __init__(self):
        self.founding_year = 0
        self.popularity = 0
        self.typization = Typization.strict
        self.is_lazy = False

    # Считывает из списка строк параметры функ.языка
    def ReadStrArray(self, strArray, index):
        # должно быт как минимум 4 непрочитанных значения в массиве
        if index >= len(strArray) - 3:
            return 0
        if not IsFactor(strArray[index]) or not strArray[index + 1].isdigit() or \
                not strArray[index + 2].isdigit() or not strArray[index + 3].isdigit():
            return 0
        self.popularity = float(strArray[index])
        self.founding_year = int(strArray[index + 1])
        int_typization = int(strArray[index + 2])
        if int_typization == 1:
            self.typization = Typization.strict
        elif int_typization == 2:
            self.typization = Typization.dynamic
        else:
            return 0
        int_lazy = int(strArray[index + 3])
        if int_lazy == 1:
            self.is_lazy = True
        elif int_lazy == 0:
            self.is_lazy = False
        else:
            return 0
        index += 4
        return index

    # Выводит свойства элемента в консоль
    def Print(self):
        print("Functional: founding year = {}, popularity = {}, typization = {}".format(self.founding_year,
                                                                                        self.popularity,
                                                                                        self.typization.name))
        pass

    # Выводит свойства элемента в файл
    def Write(self, ostream):
        ostream.write(
            "Functional: founding year = {}, popularity = {}, typization = {}\nFounding year to popularity "
            "proportion = {}".format \
                (self.founding_year, self.popularity, self.typization.name, self.Proportion()))
        pass

    # Считает отношение года создания к популярности
    def Proportion(self):
        return self.founding_year / self.popularity
